### PR TITLE
Include variable verifications in checking

### PR DESCRIPTION
**Pull request by bjornbm**
Applies to `--only-check and `--emit-lua` flags.

Fixes the observation in https://github.com/pallene-lang/pallene/issues/567#issuecomment-1479756211, and also improves on #566 which also didn't do full type checking.

There are some repeated compilation steps performed now by `--emit-lua`. A refactor allowing to resume compilation midway by passing the intermediate `prog_ast` could avoid this, but I'm not convinced the complexity would be worth it just for the sake of --emit-lua.

### CODE REVIEW

File: spec/translator_spec.lua [hunk 1]
```diff
@@ -483,6 +483,7 @@ local m: module = {}
 local xs: {any} = {10, "hello", 3.14}
 
 local function f(x: any, y: any): any
+    return nil as nil
```
Comment (hugomg) [inline spec/translator_spec.lua:486]: Could this be just `return nil`, without the `as`?

**Review by hugomg — CHANGES_REQUESTED** [review]
Thanks!

We should probably also fix the root cause of this: the `checker.lua` module and the `"checker"` step in the compilation pipeline have an outdated name. They are no longer the only place doing compile-time checking. What do you think about renaming them to or `typechecker` or `typer`?

**Comment by bjornbm** [comment]
I think that is a good idea. I was thinking the same, that the names are confusing.

I will admit that I didn't look in detail, but my impression is that `checker` does syntax checking while `uninitialized` does the type checking. So `syntaxchecker` and `typechecker` could be better names. But there may be more going on in either or both.

File: spec/translator_spec.lua (hunk 1 above)
Comment (bjornbm) [reply spec/translator_spec.lua:486]: It could be, but I added `as nil` to increase the testing of this test case. As the removal of `as …` syntax was not tested before.

Of course, I believe I could remove this line and the function return type to be closer to the original test case, which would also remove the trailing white space but arguably isn't testing much (such as the type at EOL corner case).

Comment (hugomg) [reply spec/translator_spec.lua:486]: A test for EOL stuff sounds like a good idea but I feel it deserves a separate testcase (`it` block with appropriate name)

Comment (bjornbm) [reply spec/translator_spec.lua]: Added testcase in c43d3d1d18f5867d2a2d8566c77aa24e7e8a0210.

**Comment by bjornbm** [comment]
I think I have addressed your requests/suggestions in this pull request, @hugomg.

File: src/pallene/parser.lua [hunk 2]
```diff
@@ -285,7 +285,7 @@ function Parser:Program()
                 exp.var._tag == "ast.Var.Name" and
                 exp.var.name == modname)
             then
-                -- The checker also needs to check that this name has not been shadowed
+                -- The type checker also needs to check that this name has not been shadowed
```
Comment (bjornbm) [inline src/pallene/parser.lua:288]: I wasn't sure about the meaning of this comment, and if it relates to typechecker or if something else was meant by “checker”?

File: src/pallene/typechecker.lua [hunk 3]
```diff
@@ -974,7 +974,7 @@ function Checker:check_exp_synthesize(exp)
         -- We check the child expression with verify instead of synthesize because Pallene cases
         -- also act as type annotations for things like empty array literals: ({} as {value}).
         -- However, this means that the call to verify almost always inserts a redundant cast node.
-        -- To keep the --dump-checker output clean, we get rid of it.  By the way, the Pallene to
+        -- To keep the --dump-typechecker output clean, we get rid of it.  By the way, the Pallene to
```
Comment (bjornbm) [inline src/pallene/typechecker.lua:977]: I didn't see if/where this flag was implemented. There was another reference to a “--dump compiler option” in a comment in driver.lua (line 51). But I don't see any --dump flags implemented in pallenec.lua. Were they replaced by `--emit-c` and `--print-ir`?

File: src/pallene/parser.lua (hunk 2 above)
Comment (hugomg) [reply src/pallene/parser.lua:288]: Good point. If we want to be picky, Scope Checking would be separate idea from Type Checking. But if you ask me it's fine to bundle it with the type checker (in our current implementation they are one and the same)

File: src/pallene/typechecker.lua (hunk 3 above)
Comment (hugomg) [reply src/pallene/typechecker.lua:977]: Nice catch. The dump flag was removed in #536. I'll fix that comment later.

**Review by hugomg — APPROVED** [review]
Excellent! Thank you.
